### PR TITLE
initial release

### PR DESCRIPTION
**Pull request by adamleff**
Initial release of the redhat_subscription_manager cookbook, providing custom resources for interacting with RHSM and Red Hat Satellite.

### CODE REVIEW

File: libraries/rhsm_register.rb [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    property :satellite_host,      kind_of: String
+    property :organization,        kind_of: String
+    property :environment,         kind_of: String
+    property :username,            kind_of: String
+    property :password,            kind_of: String
+    property :auto_attach,         kind_of: [ TrueClass, FalseClass ], default: false
+
+    action :register do
+      remote_file '/tmp/katello-package.rpm' do
+        source "http://#{satellite_host}/pub/katello-ca-consumer-latest.noarch.rpm"
```
Comment (edolnx) [inline libraries/rhsm_register.rb:36]: This may be a dumb question, but would we want to make this source URL a attribute for some futureproofing?

**Comment by edolnx** [comment]
Other than my comment above (which is more an aside discussion rather than a blocker) :+1:

File: libraries/rhsm_register.rb (hunk 1 above)
Comment (adamleff) [inline libraries/rhsm_register.rb:36]: It's a good question, but according to the Red Hat Satellite user guide, it's not a configurable URL, and this URL should always point to the latest katello CA available. 

My vote is to leave it as-is for now, but if it comes up with a use case, I'll definitely change it once I understand it more.  Fair?

Comment (edolnx) [inline libraries/rhsm_register.rb:36]: Works for me :shipit:

File: libraries/helpers.rb [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+module RhsmCookbook
+  module RhsmHelpers
+    def register_command
+      command = [ 'subscription-manager register ']
```
Comment (ryancragun) [inline libraries/helpers.rb]: Extra space after register.  Also `%w()` :)

Comment (adamleff) [inline libraries/helpers.rb]: Thanks!  Extra space was there when I was concatenating strings vs joining array items.  Fixing...

File: libraries/helpers.rb [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+      end
+
+      command << '--auto-attach' if auto_attach
+      command.join(' ')
+    end
+
+    def activation_keys
+      return unless activation_key
+
+      activation_key.respond_to?(:each) ? activation_key : [ activation_key ]
```
Comment (ryancragun) [inline libraries/helpers.rb]: I think you can probably change this whole method to `Array(activation_key)`

File: libraries/helpers.rb [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    def subscription_attached?(subscription)
+      cmd = Mixlib::ShellOut.new("subscription-manager list --consumed | grep #{subscription}")
+      cmd.run_command
+      !cmd.stdout.match(/Pool ID:\s+#{subscription}$/).nil?
+    end
+
+    def repo_enabled?(repo)
+      cmd = Mixlib::ShellOut.new('subscription-manager repos --list-enabled')
+      cmd.run_command
+      !cmd.stdout.match(/Repo ID:\s+#{repo}$/).nil?
```
Comment (ryancragun) [inline libraries/helpers.rb:68]: Why the double negative?

Comment (adamleff) [inline libraries/helpers.rb:68]: just forcing to be true/false vs. MatchData/NilClass.

File: libraries/helpers.rb [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    end
+
+    def serials_by_pool
+      serials = {}
+      pool = nil
+      serial = nil
+
+      cmd = Mixlib::ShellOut.new('subscription-manager list --consumed')
+      cmd.run_command
+      cmd.stdout.split("\n").each do |line|
```
Comment (ryancragun) [inline libraries/helpers.rb]: You can use `#lines` here

File: libraries/rhsm_errata_level.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+module RhsmCookbook
+  class RhsmErrataLevel < ChefCompat::Resource
+    include RhsmCookbook::RhsmHelpers
+
+    resource_name :rhsm_errata_level
+
+    property :errata_level, kind_of: String, name_property: true
+
+    action :install do
+      validate_errata_level!(errata_level)
```
Comment (ryancragun) [inline libraries/rhsm_errata_level.rb:28]: Totally OT, but are errata levels case sensitive?  If so, lame, but I'd love to see us support lowercase goodness.  If not, lets make that helper downcase all the things.

**Comment by ryancragun** [comment]
:rocket: :guitar:

File: libraries/rhsm_errata_level.rb (hunk 6 above)
Comment (adamleff) [inline libraries/rhsm_errata_level.rb:28]: Yup it's lame. I will make t more delightful though. I'll downcast everything on the way in and then provide a helper method to translate it into the appropriate case for yum to honor it. Will do that tomorrow!

**Pull request merged**